stub/src/main/java/io/grpc/stub/BlockingClientCall.java

```diff
@@ -214,27 +217,41 @@ public boolean write(ReqT request, long timeout, TimeUnit unit)
     return write(false, request, timeout, unit);
   }
 
+  /**
+   * Run inside a synchronized block so that cancel and halfClose can't run between checking for
+   * closeLock and call.sendMessage().  Can't hold the closeLock while waiting for ready as
+   * otherwise we won't be able to do halfClose or cancel from the client side while waiting.
+   * Initial check doesn't need to be in sync block since checked later, but makes tsan happy.
+   */
   private boolean write(boolean waitForever, ReqT request, long timeout, TimeUnit unit)
       throws InterruptedException, TimeoutException, StatusException {
 
-    if (writeClosed) {
-      throw new IllegalStateException("Writes cannot be done after calling halfClose or cancel");
+    // Optimization to avoid waiting for ready if writes are already closed
+    synchronized (closeLock) {
+      if (writeClosed) {
+        throw new IllegalStateException("Writes cannot be done after calling halfClose or cancel");
+      }
     }
-
     long end = System.nanoTime() + unit.toNanos(timeout);
 
     Predicate<BlockingClientCall<ReqT, RespT>> predicate =
         (x) -> x.call.isReady() || x.closedStatus != null;
     executor.waitAndDrainWithTimeout(waitForever, end, predicate, this);
-    Status savedClosedStatus = closedStatus;
-    if (savedClosedStatus == null) {
-      call.sendMessage(request);
-      return true;
-    } else if (savedClosedStatus.isOk()) {
-      return false;
-    } else {
-      // Propagate any errors returned from the server
-      throw savedClosedStatus.asException();
+    synchronized (closeLock) {
+      Status savedClosedStatus = closedStatus;
+      if (savedClosedStatus == null && writeClosed) {
+        throw new IllegalStateException("Writes cannot be done after calling halfClose or cancel");
+      }
+
+      if (savedClosedStatus == null) {
+        call.sendMessage(request);
```

Comment:
We don't want to be holding a lock while calling arbitrary user code. This is asking for deadlocks.